File: cmangos/api/account/invite.go
package account

import (
  "database/sql"

  "github.com/google/uuid"

  "metagit.org/blizzlike/cmangos-api/modules/database"
)

type InviteInfo struct {
  Token string `json:"token,omitempty"`
}

func AddAccountToInviteToken(token string, id int64) error {
  stmt, err := database.Api.Prepare(
    "UPDATE invitetoken SET account = ? WHERE token = ?;")
  if err != nil {
    return err
  }
  defer stmt.Close()

  _, err = stmt.Exec(id, token)
  if err != nil {
    return err
  }

  return nil
}

func GetInviteTokens(id int) ([]InviteInfo, error) {
  var ii []InviteInfo
  stmt, err := database.Api.Prepare(
    "SELECT token FROM invitetoken WHERE account IS NULL AND friend = ?;")
  if err != nil {
    return ii, err
  }
  defer stmt.Close()

  var rows *sql.Rows
  rows, err = stmt.Query(id)
  for rows.Next() {
    var t InviteInfo
    err = rows.Scan(&t.Token)
    if err != nil {
      return ii, err
    }

    ii = append(ii, t)
  }

  return ii, nil
}

func WriteInviteToken(id int) (string, error) {
  t, _ := uuid.NewRandom()
  token := t.String()
  stmt, err := database.Api.Prepare(
    "INSERT INTO invitetoken (token, friend) VALUES (?, ?);")
  if err != nil {
    return token, err
  }
  defer stmt.Close()

  _, err = stmt.Exec(token, id)
  if err != nil {
    return token, err
  }

  return token, nil
}

File: lua/cmangos/realmd/account/invite.lua
local uuid = require("uuid")

local _M = {}

function _M.create_token(self, id, gmlevel, info)
  local token = uuid.create()
  local result, err = sql.api.query(
    "INSERT INTO invitetoken (token, friend, gmlevel, info) VALUES (?, ?, ?, ?)",
    token, id, gmlevel, info or "")
  if err then return nil, err end

  return token
end

function _M.delete_token(self, token)
  local _, err = sql.api.query(
    "DELETE FROM invitetoken WHERE token = ? AND account IS NULL;", token)
  if err then return nil, err end

  return true
end

function _M.get_tokens(self, id)
  local result, err = sql.api.query(
    "SELECT * FROM invitetoken WHERE friend = ? AND account IS NULL;", id)
  if err then return nil, err end
  return result
end

function _M.update_token(self, token, id)
  local result, err = sql.api.query(
    "UPDATE invitetoken SET account = ? WHERE token = ?;",
    id, token)
  if err ~= nil then return nil, err end

  return true
end

function _M.validate_token(self, token)
  local result, err = sql.api.query(
    "SELECT friend, token, gmlevel FROM invitetoken " ..
    " WHERE token = ? AND account IS NULL;", token)
  if err then return nil, err end
  if #result == 0 then return nil end

  return result[1]
end

return _M
